javalib/src/main/scala/java/lang/ref/ReferenceQueue.scala

```diff
@@ -4,16 +4,35 @@ import scalanative.annotation.stub
 import scala.collection.mutable
 
 class ReferenceQueue[T >: Null <: AnyRef] {
-  private val underlying = mutable.Queue[Reference[_]]()
-  private[java] def add(reference: Reference[_]): Unit =
-    underlying += reference
+  private val underlying = mutable.Queue[Reference[_ <: T]]()
+  private[ref] def enqueue(reference: Reference[_ <: T]): Unit =
+    synchronized {
+      underlying += reference
+      notify()
+    }
 
-  def poll(): java.lang.ref.Reference[_] =
-    underlying.dequeueFirst(ref => true).getOrElse(null)
+  def poll(): Reference[_ <: T] = {
+    synchronized[Reference[_ <: T]] {
+      underlying
+        .dequeueFirst(_ => true)
+        .map(_.dequeue())
+        .orNull
+    }
+  }
 
-  @stub
-  def remove(): java.lang.ref.Reference[_] = ???
+  def remove(): Reference[_ <: T] =
+    remove(0)
 
-  @stub
-  def remove(timeOut: Long): java.lang.ref.Reference[_] = ???
+  def remove(timeOut: Long): Reference[_ <: T] = {
+    if (timeOut < 0) throw new IllegalArgumentException()
+
+    synchronized[Reference[_ <: T]] {
+      if (timeOut == 0) {
+        while (underlying.isEmpty) wait()
+      } else {
+        if (underlying.isEmpty) wait(timeOut)
+      }
```

Comment:
[It is recommended](https://docs.oracle.com/javase/10/docs/api/java/lang/Object.html#wait) to guard wait() inside while loop and recompute timeout to prevent issues with spurious wakeups of the thread. Also `wait()` is equal to `wait(0L)` or `wait(0L, 0)` so it can be safely used instead.   
```suggestion
      def now() = System.currentTimeMillis()
      val deadline = now() + timeout
      def timeoutExceeded(current: Long): Boolean = {
        if(timeout == 0) false
        else current > deadline 
      }
      
      while(underlying.isEmpty && !timeoutExceeded(now())) {
        val timeoutMillis = (deadline - now()).min(0L)
        wait(timeoutMillis)
      }
```